core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -456,9 +460,16 @@ class RequestChannel(val queueSize: Int,
     }
   }
 
-  /** Get the next request or block until specified time has elapsed */
-  def receiveRequest(timeout: Long): RequestChannel.BaseRequest =
-    requestQueue.poll(timeout, TimeUnit.MILLISECONDS)
+  /** Get the next request or block until specified time has elapsed 
+   *  Check the callback queue and execute first if present since these
+   *  requests have already waited in line. */
+  def receiveRequest(timeout: Long): RequestChannel.BaseRequest = {
+    val callbackRequest = callbackQueue.poll()
+    if (callbackRequest != null)
+      callbackRequest
+    else 
+      requestQueue.poll(timeout, TimeUnit.MILLISECONDS)
```

Comment:
We could probably handle WakeupRequest in this function, so that the wakeup mechanism is encapsulated in RequestChannel (i.e. check if we got a wakeup request from the requestQueue and poll the callbackQueue again in that case).